### DISCUSSION TITLE
no ESP32 support?

### DISCUSSION THREAD

**Discussion by carloslema** [forntoh/LcdMenu#11] [Ideas]
Hi Forntoh, thank you so much for this great library.

I am trying to get it to work with an ESP32 and 20x4 LCD.

Unfortunately, with the Basic example, it firesd a compiling Error with int the Arduino IDE.

While you look into that, Is there a way to drawMenu on an Event — for example pressing on an Encoder Switch?
This bring me to my next question, I see that there is a setText(), since LcdMenu takes over the LCD, is there a way to print to the LCD other information that is not menu related?

Scenario: The device is displaying some information. User clicks on an encoder switch, the menu shows up — user selects an action, and when it clicks on exit, back to the main screen fully exiting the menu.

Thanks in advance for your help

**Comment by forntoh** [comment]
Concerning your question about **ESP43 support,** if you could dump the compile logs here, it would help a lot. I haven't yet tested if the library works _out-of-the-box_ with the ESP, I can add support if that's not the case.

_Thanks, for your feedback, the library has become 1% better because of it_ 😜

**Reply by carloslema** [reply]
Thank you so much for your blazing fast response.

---- the actual code I am testing — removed anything related to keypad to isolate the problem
For what is worth, I am using Arduino IDE. I ran into a bunch of issues with Atom + Platformio. I will not mess up my VScode invironement with Platformio.

I manually downloaded your LcdMenu-master libary an dplaced in the Arduino Library folder.

`#include <LcdMenu.h>
LcdMenu menu(4, 20); // Constructor
extern MenuItem mainMenu[]; // define

// Init menu items
MenuItem mainMenu[] = {ItemHeader(),
                       MenuItem("Start service"),
                       MenuItem("Connect to WiFi"),
                       MenuItem("Settings"),
                       MenuItem("Blink SOS"),
                       MenuItem("Blink random"),
                       ItemFooter()};

void setup() {
    menu.setupLcdWithMenu(0x27, mainMenu); // init
}
void loop() {    
}`

Comments
The first waring I ignore since it works fine with other sketches.

compiler errors:
`

In file included from ~/Arduino/ESP32_lcdMenu/ESP32_lcdMenu.ino:10:0:

~/Arduino/libraries/LcdMenu-master/src/LcdMenu.h:270:5: error: extra qualification 'LcdMenu::' on member 'LcdMenu' [-fpermissive]
     LcdMenu::LcdMenu(uint8_t maxRows, uint8_t maxCols) {
     ^
~/Arduino/libraries/LcdMenu-master/src/LcdMenu.h:283:10: error: extra qualification 'LcdMenu::' on member 'setupLcdWithMenu' [-fpermissive]
     void LcdMenu::setupLcdWithMenu(
          ^
~/Arduino/libraries/LcdMenu-master/src/LcdMenu.h: In member function 'void LcdMenu::setupLcdWithMenu(uint8_t, MenuItem*)':
~/Arduino/libraries/LcdMenu-master/src/LcdMenu.h:287:14: error: 'class LiquidCrystal_I2C' has no member named 'init'
         lcd->init();
              ^
Multiple libraries were found for "LcdMenu.h"
 Used: ~/Arduino/libraries/LcdMenu-master
Multiple libraries were found for "LiquidCrystal_I2C.h"
 Used: ~/Arduino/libraries/Arduino-LiquidCrystal-I2CE
Multiple libraries were found for "Wire.h"
 Used: ~/Library/Arduino15/packages/esp32/hardware/esp32/1.0.4/libraries/Wire
exit status 1
Error compiling for board Adafruit ESP32 Feather.
`

**Reply by forntoh** [reply]
This library is designed out of the base to support I2C _(easy to connect/configure),_ If you're not using I2C, turn it off by adding this line below your imports
```cpp
#define USE_STANDARD_LCD
```

> From my testing, I don't think the ESP supports communication with an LCD through the I2C protocol

**Reply by carloslema** [reply]
Yup. Saw that too — line 31. I am using the <LiquidCrystal_I2C.h> without any problems. At one point I even deleted anything related to <LiquidCrystal.h> and 259, that's when I got only one error...
~/Arduino/libraries/LcdMenu-master/src/LcdMenu.h:270:5: error: extra qualification 'LcdMenu::' on member 'LcdMenu' [-fpermissive]

But couldn't figure out why that is hapening... in the skecth, LcdMenu menu(4, 20) creates the menu object. then you build the menu array (mainMenu), and attach to the menu via menu.setupLcdWithMenu(0x27, mainMenu);

I don't know what " extra qualification" or [-fpermissive] means — LcdMenu.h:270:5

**Reply by forntoh** [reply]
I'll check that

**Reply by carloslema** [reply]
@forntoh quick update. I got it to compile on the ESP32.

What I did.
I deleted  LcdMenu:: in the constructor and the setupLcdWithMenu() function and it works!

Now that I got it to work, I am looking into using it with an Encoder that has a switch. 

Thanks again for such a great library and such a stright forward implementation.

\\ carlos

**Reply by forntoh** [reply]
Okay, thanks for promoting this library as a user and also for providing feedback that improves the library for future and current users.

I will also try removing `LcdMenu::` and test it on the Uno, if it works, I'll patch the current release.

**Reply by forntoh** [reply]
Oh... it wasn't even supposed to be there, I moved all the code from the `.cpp` file to the header file to permit `NO I2C` functionality using `#define`, I just forgot to clean up the code properly 😖

**Reply by carloslema** [reply]
So far the test prototype that uses an encoder is working 75% — I'll submit a pull request so you can take a look at the small changes I've made and a simple example.

But I am stuck, and this is very silly. I have not figured out how to remove the menu once is created. I can initialize it with a switch (splitting LCD int & creating the menu obj.)

I tried a few things: from simple lcd->clear(), MenuItem  noMenu[] = {} to brute force by deleteting the pointer to the menu (delete[] this->currentMenuTable; / delete &this->currentMenuTable;) bad idea (crahsed).

The goal is to display and hide the menu system at will — safely.

Again, thanks for your help!

**Reply by forntoh** [reply]
Ok, I will check your PR, I don't really have time to improve the library at the moment, I'll only be able to do that next year, be expecting loads of features when I start 😎.

I am noting all the problems that you're facing and visualizing how I can fix them, but unfortunately, I can't code any time soon. I open to PRs, I'm waiting for your PRs

**Comment by forntoh** [comment]
> While you look into that, Is there a way to drawMenu on an Event — for example pressing on an Encoder Switch?

Can you explain how the Encoder **switch press event** is triggered? 🤔 I will know how to proceed from there.

Is this feature similar to #9

**Reply by carloslema** [reply]
Saw issue #9 — It seems to be more related to a "variable status update" that gets printed to a menu item.

My goal is to use an encoder that has a switch. When the device is powered, it displays text on the lcd. When the encoder is pressed, it goes into an 'Edit Mode.'  When a user rotates the encoder, it moves the cursor up or down. If the user presses again on the encoder, it selects that menu item. Either a switch statement or FSM might help sort this out.

**Comment by forntoh** [comment]
> is there a way to print to the LCD other information that is not menu related?

There's a feature similar to this, check it out [here](https://github.com/forntoh/LcdMenu/wiki/Example-05-Menu-Notifications)

> Scenario: The device is displaying some information. User clicks on an encoder switch, the menu shows up — user selects an action, and when it clicks on exit, back to the main screen fully exiting the menu.

Sounds interesting, I can build it up on the feature mentioned above 👆

**Reply by carloslema** [reply]
Yes, I saw that. Time-based Notifications. 
In the example you have, it does not perform any password verification logic — which is another thing in itself. It just waits 2sec and print the message.

But, conceptually you are printing to the screen. In looking through LcdMenu.h, the method/function on line 513 can be used to create a print to screen function, something crude like:

void displayMessage((char* message, uint8_t setCursorX, uint8_t setCursorY)){
lcd.clear();
lcd.setCursor(setCursorX, setCursorY);
lcd.print(message);
}
This function does not need to know where the cursor is, it just clears the screen and print the text on what ever line. This way I can use it to display information on the LCD when the menu is not engaged.

**Reply by forntoh** [reply]
### Summary 
Okay, so the library should provide these when a callback is executed: 
- A method to display just text _(user can specify position on screen, or leave it)_
- Show a temporary menu with _callback-bound_ options, whereby the temporary menu disappears when any option is selected _(enter)_.

**Reply by carloslema** [reply]
you are super fast.... thank you so much for looking into this!

I don't know if these need to be calback per se.  I guess the exit "Edit Mode" is a callback... but dunno what you mean.

"Show a menu with actions that disappear when the actions are executed."
Mmm... dunno what you mean by this.

A very good example is a clock. The clock displays the time indefinitely. But, if the user clicks on the "edit time", the clock goes in to "edit time" mode — and clock displays a couple of menu options — Set Time, set alarm etc. The user select "Set Time." User select hours and/or minutes and set them. Then the user clicks again to "set/update time" and exists when done. Now Clock is back to displaying the correct time.

Let me create a logic diagram to illustrate what I mean.

**Reply by forntoh** [reply]
That exactly what I mean, you have the primary menu showing, then some action triggers, an overly menu is shown, once any option on this menu is clicked, the action bound to the option is executed and the display reverts back to the primary menu.

**Reply by carloslema** [reply]
Awesome!
